storm-core/src/jvm/org/apache/storm/stats/BoltExecutorStats.java

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.stats;
+
+import clojure.lang.PersistentVector;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.storm.metric.internal.MultiCountStatAndMetric;
+import org.apache.storm.metric.internal.MultiLatencyStatAndMetric;
+
+@SuppressWarnings("unchecked")
+public class BoltExecutorStats extends CommonStats {
+
+    public static final String ACKED = "acked";
+    public static final String FAILED = "failed";
+    public static final String EXECUTED = "executed";
+    public static final String PROCESS_LATENCIES = "process-latencies";
+    public static final String EXECUTE_LATENCIES = "execute-latencies";
+
+    public static final String[] BOLT_FIELDS = {ACKED, FAILED, EXECUTED, PROCESS_LATENCIES, EXECUTE_LATENCIES};
+
+    public BoltExecutorStats() {
+        super();
+
+        put(ACKED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(FAILED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(EXECUTED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(PROCESS_LATENCIES, new MultiLatencyStatAndMetric(NUM_STAT_BUCKETS));
+        put(EXECUTE_LATENCIES, new MultiLatencyStatAndMetric(NUM_STAT_BUCKETS));
+    }
+
+    public MultiCountStatAndMetric getAcked() {
+        return (MultiCountStatAndMetric) this.get(ACKED);
+    }
+
+    public MultiCountStatAndMetric getFailed() {
+        return (MultiCountStatAndMetric) this.get(FAILED);
+    }
+
+    public MultiCountStatAndMetric getExecuted() {
+        return (MultiCountStatAndMetric) this.get(EXECUTED);
+    }
+
+    public MultiLatencyStatAndMetric getProcessLatencies() {
+        return (MultiLatencyStatAndMetric) this.get(PROCESS_LATENCIES);
+    }
+
+    public MultiLatencyStatAndMetric getExecuteLatencies() {
+        return (MultiLatencyStatAndMetric) this.get(EXECUTE_LATENCIES);
+    }
+
+    public void boltExecuteTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getExecuted().incBy(key, this.rate);
+        this.getExecuteLatencies().record(key, latencyMs);
+    }
+
+    public void boltAckedTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getAcked().incBy(key, this.rate);
+        this.getProcessLatencies().record(key, latencyMs);
+    }
+
+    public void boltFailedTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getFailed().incBy(key, this.rate);
+
+    }
+
+    public Map renderStats() {
+        cleanupStats();
+        Map ret = new HashMap();
+        ret.putAll(valueStats(CommonStats.COMMON_FIELDS));
+        ret.putAll(valueStats(BoltExecutorStats.BOLT_FIELDS));
+        StatsUtil.putRawKV(ret, StatsUtil.TYPE, StatsUtil.KW_BOLT);
```

Comment:
@revans2 I've changed all the others according to your comments, my only problem is renderStats, do you mean that I only need to maintain ExecutorStats in memory without any stats Maps while running? Although renderStats is only used in thriftify-zk-worker-hb, when workers do heartbeats(see worker/do-executor-heartbeats method), it will use a clojure PersistentMap data structure, which messes with ExecutorStats, should I turn it into ExecutorStats too?